network/msgCompressor.go

```diff
@@ -0,0 +1,151 @@
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/DataDog/zstd"
+
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+var zstdCompressionMagic = [4]byte{0x28, 0xb5, 0x2f, 0xfd}
+
+// checkCanCompress checks if there is an proposal payload message and peers supporting compression
+func checkCanCompress(request broadcastRequest, prio bool, peers []*wsPeer) bool {
+	canCompress := false
+	if prio {
+		hasPP := false
+		for _, tag := range request.tags {
+			if tag == protocol.ProposalPayloadTag {
+				hasPP = true
+				break
+			}
+		}
+		// if have proposal payload check if there are any peers supporting compression
+		if hasPP {
+			for _, peer := range peers {
+				if peer.vfCompressedProposalSupported() {
+					canCompress = true
+					break
+				}
+			}
+		}
+	}
+	return canCompress
+}
+
+// zstdCompressMsg returns a concatenation of a tag and compressed data
+func zstdCompressMsg(tbytes []byte, d []byte) ([]byte, string) {
+	bound := zstd.CompressBound(len(d))
+	mbytesComp := make([]byte, len(tbytes)+bound)
+	copy(mbytesComp, tbytes)
+	comp, err := zstd.Compress(mbytesComp[len(tbytes):], d)
+	if err != nil {
+		// fallback and reuse non-compressed original data
+		logMsg := fmt.Sprintf("failed to compress into buffer of len %d: %v", len(d), err)
+		copied := copy(mbytesComp[len(tbytes):], d)
+		return mbytesComp[:len(tbytes)+copied], logMsg
+	}
+	mbytesComp = mbytesComp[:len(tbytes)+len(comp)]
+	return mbytesComp, ""
+}
+
+// MaxDecompressedMessageSize defines a maximum decompressed data size
+// to prevent zip bombs
+const MaxDecompressedMessageSize = 20 * 1024 * 1024 // some large enough value
+
+// wsPeerMsgDataConverter performs optional incoming messages conversion.
+// At the moment it only supports zstd decompression for payload proposal
+type wsPeerMsgDataConverter struct {
+	log    logging.Logger
+	origin string
+
+	// actual converter(s)
+	ppdec zstdProposalDecompressor
+}
+
+type zstdProposalDecompressor struct {
+	active bool
+}
+
+func (dec zstdProposalDecompressor) enabled() bool {
+	return dec.active
+}
+
+func (dec zstdProposalDecompressor) accept(data []byte) bool {
+	return len(data) > 4 && bytes.Equal(data[:4], zstdCompressionMagic[:])
+}
+
+func (dec zstdProposalDecompressor) convert(data []byte) ([]byte, error) {
+	r := zstd.NewReader(bytes.NewReader(data))
+	defer r.Close()
+	b := make([]byte, 0, 1024)
```

Comment:
I though we are only compressing large messages, so maybe start higher?  We could define a constant `minCompressedMsgSize`?

Reply:
we compressing all proposals, small and large. I think I like this idea, the magic check above would handle both compressed and non-compressed

Reply:
oh, right, mspacked, my bad

Reply:
I was considering suggesting that too.  Then you could io.ReadFull on the known size (allocated all at once). And the check for exploding messages isn't even needed (it'll just fail to decompress into the buffer, which presumably you've prechecked against the maxsize)

Reply:
Or, how about we start with a buffer 5x the size of the message?

Reply:
maybe 2.5x would be a good guess since we see 2-3x compression ratio

Reply:
set to 3x

Reply:
Sounds good.  That should eliminate most copies.